File: iOS App/VoiceBlogger/VoiceBlogger/State/AppState.swift
import Foundation
import Observation
import SwiftData

enum AppStage: Equatable {
    case modelDownload
    case recording
    case transcribing(post: BlogPost)
    case preparingBlog(postID: UUID)
    case generatingBlog(post: BlogPost)
    case viewingBlog(post: BlogPost)
    case viewingInstagram(post: BlogPost)
    case history

    static func == (lhs: AppStage, rhs: AppStage) -> Bool {
        switch (lhs, rhs) {
        case (.modelDownload, .modelDownload): return true
        case (.recording, .recording): return true
        case (.history, .history): return true
        case (.transcribing(let a), .transcribing(let b)): return a.id == b.id
        case (.preparingBlog(let a), .preparingBlog(let b)): return a == b
        case (.generatingBlog(let a), .generatingBlog(let b)): return a.id == b.id
        case (.viewingBlog(let a), .viewingBlog(let b)): return a.id == b.id
        case (.viewingInstagram(let a), .viewingInstagram(let b)): return a.id == b.id
        default: return false
        }
    }
}

@Observable
final class AppState {
    var stage: AppStage = .recording
    var errorMessage: String?
    var showError = false
    @ObservationIgnored var generationModelContext: ModelContext?

    func navigateTo(_ stage: AppStage) {
        self.stage = stage
    }

    func showError(_ message: String) {
        errorMessage = message
        showError = true
    }

    func dismissError() {
        showError = false
        errorMessage = nil
    }
}

File: Android-App/androidApp/src/main/kotlin/com/anup/voiceblogger/state/AppState.kt
package com.anup.voiceblogger.state

import androidx.lifecycle.ViewModel
import com.anup.voiceblogger.data.BlogPost
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

sealed class AppStage {
    object ModelDownload : AppStage()
    object Onboarding : AppStage()
    object Recording : AppStage()
    data class Transcribing(val post: BlogPost) : AppStage()
    data class PreparingBlog(val postId: String) : AppStage()
    data class GeneratingBlog(val post: BlogPost) : AppStage()
    data class ViewingBlog(val post: BlogPost) : AppStage()
    data class ViewingInstagram(val post: BlogPost) : AppStage()
    data class ViewingLinkedIn(val post: BlogPost) : AppStage()
    object History : AppStage()
    object About : AppStage()
}

class AppStateViewModel : ViewModel() {
    private val _stage = MutableStateFlow<AppStage>(AppStage.ModelDownload)
    val stage: StateFlow<AppStage> = _stage.asStateFlow()

    fun navigateTo(stage: AppStage) {
        _stage.value = stage
    }

    fun goToRecording() {
        _stage.value = AppStage.Recording
    }

    fun goToTranscribing(post: BlogPost) {
        _stage.value = AppStage.Transcribing(post)
    }

    fun goToPreparingBlog(postId: String) {
        _stage.value = AppStage.PreparingBlog(postId)
    }

    fun goToGeneratingBlog(post: BlogPost) {
        _stage.value = AppStage.GeneratingBlog(post)
    }

    fun goToViewingBlog(post: BlogPost) {
        _stage.value = AppStage.ViewingBlog(post)
    }

    fun goToViewingInstagram(post: BlogPost) {
        _stage.value = AppStage.ViewingInstagram(post)
    }

    fun goToViewingLinkedIn(post: BlogPost) {
        _stage.value = AppStage.ViewingLinkedIn(post)
    }

    fun goToHistory() {
        _stage.value = AppStage.History
    }
}
